sklearn/tree/_criterion.pyx

```diff
@@ -409,79 +396,66 @@ cdef class ClassificationCriterion(Criterion):
             child to the left child.
         """
         cdef DOUBLE_t* y = self.y
-        cdef SIZE_t y_stride = self.y_stride
-        cdef DOUBLE_t* sample_weight = self.sample_weight
-
-        cdef SIZE_t* samples = self.samples
         cdef SIZE_t pos = self.pos
         cdef SIZE_t end = self.end
 
-        cdef SIZE_t n_outputs = self.n_outputs
-        cdef SIZE_t* n_classes = self.n_classes
-        cdef SIZE_t label_count_stride = self.label_count_stride
-        cdef double* label_count_left = self.label_count_left
-        cdef double* label_count_right = self.label_count_right
-        cdef double* label_count_total = self.label_count_total
+        cdef double* node_sum_left = self.node_sum_left
+        cdef double* node_sum_right = self.node_sum_right
+        cdef double* node_sum = self.node_sum
 
-        cdef SIZE_t i = 0
-        cdef SIZE_t p = 0
-        cdef SIZE_t k = 0
-        cdef SIZE_t c = 0
-        cdef SIZE_t label_index = 0
+        cdef SIZE_t i, p, k, c
+        cdef SIZE_t label_index
         cdef DOUBLE_t w = 1.0
-        cdef DOUBLE_t diff_w = 0.0
 
         # Update statistics up to new_pos
         #
         # Given that
-        #   label_count_left[x] +  label_count_right[x] = label_count_total[x]
-        # and that label_count_total is known, we are going to update
-        # label_count_left from the direction that require the least amount
+        #   node_sum_left[x] +  node_sum_right[x] = node_sum[x]
+        # and that node_sum is known, we are going to update
+        # node_sum_left from the direction that require the least amount
         # of computations, i.e. from pos to new_pos or from end to new_po.
 
         if (new_pos - pos) <= (end - new_pos):
             for p in range(pos, new_pos):
-                i = samples[p]
+                i = self.samples[p]
 
-                if sample_weight != NULL:
-                    w = sample_weight[i]
+                if self.sample_weight != NULL:
+                    w = self.sample_weight[i]
 
-                for k in range(n_outputs):
-                    label_index = (k * label_count_stride +
-                                   <SIZE_t> y[i * y_stride + k])
-                    label_count_left[label_index] += w
+                for k in range(self.n_outputs):
+                    label_index = (k * self.label_count_stride +
+                                   <SIZE_t> y[i * self.y_stride + k])
+                    node_sum_left[label_index] += w
 
-                diff_w += w
+                self.weighted_n_left += w
+                self.weighted_n_right -= w
```

Comment:
I understand. I was concerned it might take more time, but it ended up not really affecting the duration while I think it is more explicit what is happening now. If you think it doesn't make the code more explicit (my intention) I can change it back.